### ISSUE TITLE
The Item Links Library

### ISSUE THREAD

**Issue by KVonGit** [ThePix/QuestJS#44]
Hello.

Sorry.  I should have thought of it but didn't until Pertex pointed out that the links only show up in the room description.

I am working on this tonight.

---
Also, I have decided that I'm probably driving you crazy, and so decided to create a repo with my current work in progress.  There, I have my own issues (literally and ...well... literally!).

If you ever get bored, I try to keep them up to date: https://github.com/KVonGit/Quest6Game1/issues

I just started doing this, but I imagine I'll answer most of my own questions this way.

---
Also also, I think (maybe) there's a way to post JS on here that allows anyone with the link to just link to it in a ```<script>```.  That might come in handy, if true.  (Just throwing crap against the wall, here.)

**Comment by KVonGit** [comment]
PS

I'm making progress.

Here's that Issue: https://github.com/KVonGit/Quest6Game1/issues/1

**Comment by ThePix** [comment]
> Sorry. I should have thought of it but didn't until Pertex pointed out that the links only show up in the room description.

I would suggest the only other place is where the author specifically wants it, as indicated by a text processor command. If you start searching through all output text for item names, you are likely to find places it should not be.

> Also also, I think (maybe) there's a way to post JS on here that allows anyone with the link to just link to it in a <script>. That might come in handy, if true. (Just throwing crap against the wall, here.)

Any code uploaded can be linked to, as long as the CSP allows it. Have to be set specifically to this repository (or whoever) to stop just any code.

The risk is that a change in the script will break a game, and at the moment such changes do happen.

**Comment by ThePix** [comment]
I am a liitle concerned about using turnscripts for this, as it then stops the author using them on that object. Could you use `io.modulesToUpdate` instead?
https://github.com/ThePix/QuestJS/wiki/Custom-UI-Elements

And I am not too sure it is updating properly. I have a soup can that can be opened (but not using the template); the side pane updates but the object link does not. It does update for TKE and DROP.

Also, how is cmdOpps used? It will make translating to other languages harder, and I cannot see why it is needed, as you are using getVerbs anyway.

**Comment by KVonGit** [comment]
> I am a liitle concerned about using turnscripts for this, as it then stops the author using them on that object. Could you use io.modulesToUpdate instead?

I was thinking I needed to figure out how to do this yesterday.  I'll read up on it and give it a go.

---
> And I am not too sure it is updating properly.

I was returning ```true``` instead of ```world.SUCCESS``` in my TAKE mod.

The version online now should handle that correctly.

The newest version of the game is on the site: http://media.textadventures.co.uk/games/R-PODbpzSk_rxIcdAq1RQA/page.html

The newest version of this library is here: https://github.com/KVonGit/Quest6Game1/blob/main/game/rhfiles/objectLinksLib.js

NOTE: The "newest" version online is now old.  I just made it iterate through items in open containers and held by npcs.

Also, I cleaned up that Issue a little. My TODO checklist is all in the first post now.

https://github.com/KVonGit/Quest6Game1/issues/1

---
> Also, how is cmdOpps used? It will make translating to other languages harder, and I cannot see why it is needed, as you are using getVerbs anyway.

That's a cheap hack.  I couldn't figure out a better way to find commands' opposites.

I may not completely understand how getVerbs works if I can use it to switch TAKE to DROP and all that without any extra code (such as cmdOpps).  cmdOpps is just a dictionary to easily access the opposites of commands.

---
By the way, I realize I may be flooding your time with this stuff . . .  People on the forum have been complaining that the Quest server is down for at least 24 hours.  You probably know that already, but I'd feel guilty if I didn't throw out the "heads up".  You mentioned taking over for manowar.  So, I'm thinking you're left dealing with everything site-related these days.

**Comment by ThePix** [comment]
> I was returning true instead of world.SUCCESS in my TAKE mod.

I do not think that was the issue I was having...

> I was thinking I needed to figure out how to do this yesterday. I'll read up on it and give it a go.

This might start you off. Create an object and send to io to be updated, then given an update junction that will be called. It can grab all the items from the web page, so it knows wat to up date (but you might want to remove the class from items later).

```
const itemLinks = {}
io.modulesToUpdate.push(itemLinks)
itemLinks.update = function() {
  const elements = document.getElementsByClassName('object-link')
...
  }
}
```

The Quest server is in the process of being passed to me, but I do not have it yet, so nothing I can do about it being down.

**Comment by KVonGit** [comment]
> The Quest server is in the process of being passed to me, but I do not have it yet, so nothing I can do about it being down.

Gotcha.  Just making sure I wasn't sitting on pertinent information.

---
I am about to attempt messing with ```io.modulesToUpdate```.  If you're religious, pray for me. Else if superstitious, cross your fingers.  Else, hope for the best.

**[ MANIACAL LAUGH ]**

**Comment by KVonGit** [comment]
> but you might want to remove the class from items later

I didn't mess with this part (yet?), and I don't see any errors or anything.

What should I be looking out for if I don't change the order in which the class gets removed?  (I'm genuinely asking.  Making it handle these links correctly is quite the learning experience.)

**Comment by KVonGit** [comment]
I think I've got it (mostly) ready.

I revamped the verb update system almost entirely.

---
Version 0.24

The newest version of the game is on the site: http://media.textadventures.co.uk/games/R-PODbpzSk_rxIcdAq1RQA/page.html

The newest version of this library is here: https://github.com/KVonGit/Quest6Game1/blob/main/game/rhfiles/objectLinksLib.js

---
PS

Check out what I did with my [landing page](https://github.com/KVonGit/Quest6Game1/blob/main/index.html).  I added two scripts from the game to it, and now I can use settings properties from the game to fill in the title, version, and all that stuff on the landing page.  (I'm a lazy cheater.)

Here it is [on textadventures](http://media.textadventures.co.uk/games/R-PODbpzSk_rxIcdAq1RQA/index.html).

I need to change the style of that landing page.  I know.  That's on the back burner.

**Comment by KVonGit** [comment]
Version 0.25 is online.  (Same links.)

I have not tested UNDO, and I am afraid to do so.

I might try it after a break, in an hour or so.

Anyway, feel free to try to break stuff or see if I did something differently than you would prefer.

**Comment by KVonGit** [comment]
> I have not tested UNDO, and I am afraid to do so.

It worked.

---
Now, I'm wondering if I should have made ```nm``` return a link?  If they are interacting with something that utilizes ```nm``` in the output, it's most probably already got a link on the screen.  (Try  U. U. OPEN FRIDGE.  Note the two "refrigerator" links.  I guess they hit the document simultaneously, therefore there is no "first" one to be deactivated by my code.)

Also, if we think ```nm``` should return a link, I should probably modify ```nms``` and ```pv``` as well; no?

You may have already addressed this, but I can't find it in any threads I've perused.  I've been in sixth gear for a few consecutive days and may have gone slightly madder than normal. (Ha ha ha!)

**Comment by ThePix** [comment]
As an aside...

> I added two scripts from the game to it, and now I can use settings properties from the game to fill in the title, version, and all that stuff on the landing page. (I'm a lazy cheater.)

This is not lazy, but very much how you should do it. It ensures that changes in one also appears in the other. If there is a spelling mistake that gets corrected, you only have to change it in one place. This is very much the preferred way to do things in coding.

**Comment by KVonGit** [comment]
Curses!

Still not ready.  I just realized I forgot to use ```lastJoiner``` when converting the lists of links into text.

After this, I just need to know if we want ```nm``` to output a link.  (I have it set up, but I'm thinking you thought it was overkill.  I'm content either way.)

**Comment by KVonGit** [comment]
Hello.

I'm making progress.

If I can figure out a way to replace the displayed alias with a link, while leaving the alias as the actual name (with no link), I think I'll have it pretty much licked -- with much less code than the previous library.

I don't think I can fool with the actual alias because ```getName``` uses it to do all the grammar stuff.  So, right now, I have a ```linkAlias``` attribute on everything, but I can't figure out how to substitute it without altering ```lang.getName```.  Wait.  I just answered myself (I think).  I'll just be sure to code the mod to check for ```settings.linksEnabled``` before performing any item-link magic.

I'm gonna go ahead and post this comment in case you'd advise handling this differently.

I'd also like to note that rewriting all this code doesn't bother me at all because:
1. I had fun and learned a lot while writing the first version.
2. I didn't really ask for much input during the process. I was just playing around and ignorant of the fact that this would be included in the next release.
3. I'm having fun and learning a lot while rewriting it.

Any ideas?  Or is there already something on the wiki that covers this?

**Comment by ThePix** [comment]
I think you already found all these is on the Wiki. I am not entirely sure quite how you are intending to do this, but remember it has to support other languages, so replacing lang.getName would be problematic.

**Comment by KVonGit** [comment]
Help me, please, Pixie.  I am stumped.

https://github.com/KVonGit/Quest6Game1/issues/1#issuecomment-748726614

If this mod of lang.getName would problematic, I can do away with that.  That is not a problem.

...but I can't fathom why the name modifier has a different output than calling the function it calls on its own.

**Comment by KVonGit** [comment]
Never mind.  Sorry.  I should have gave myself 5 more minutes before asking.

```js
createItem("table", SURFACE(), {
	loc:"cellar",
	examine:(...params)=>{
		msg("A wooden table.");
		handleExamineHolder(params)
	}
})
```

![image](https://user-images.githubusercontent.com/30656341/102734948-c641d500-4306-11eb-9025-f9ebb5c032b0.png)

**Comment by ThePix** [comment]
> Never mind. Sorry. I should have gave myself 5 more minutes before asking.

I often find the process of asking the question makes me think of other things to check or try.

So is this nearly done, do you think?

**Comment by KVonGit** [comment]
> So is this nearly done, do you think?

I believe so.  It will hopefully be done by this evening.

1. I currently have ```lang.getName``` modified to make it work, but I think I just found a way around that (maybe).
    - https://github.com/KVonGit/Quest6Game1/blob/main/game/rhfiles/item_links.js#L469
2. I think I need to replace (or heavily modify at least one of) these two crazy functions:
    - https://github.com/KVonGit/Quest6Game1/blob/main/game/rhfiles/item_links.js#L223
    - https://github.com/KVonGit/Quest6Game1/blob/main/game/rhfiles/item_links.js#L265
3. I still feel like I could pull all of this off with much less code.

---
I just updated the files on GitHub and the game on the textadventures site.

I've changed a lot since the last time you tested anything.

http://media.textadventures.co.uk/games/R-PODbpzSk_rxIcdAq1RQA/page.html

**Comment by KVonGit** [comment]
If you're waiting on this library for the next version release of Quest, my advice would be to remove the current file and release without this.

Before I say this, I'm being negative or emotional or anything.  This is not a complaint or me looking for any pats on the back; it's just the truth:  I've been spending nearly every waking moment working on these item links for an unknown amount of days.  So, this isn't not ready because I just need to buckle down and finish. This is not ready because I am missing a fundamental building block (or two).

I really want to get rid of the ```linkStringer``` and ```createChildrenLinkString``` functions (which work together).  I honestly have no clue how I made that work at all, and I don't think it adds a last joiner except at the very end.  I have to add a few more containers to the first room to test this theory, but I bet the theory is proven quickly.

In place of those two functions, I think I just need to steal the code from the default name modifier for containers.  I think that checks contents recursively, but I'm not certain.  If it does not check contents recursively, I just need to modify the ```createChildrenLinkString``` code so it pushes to an array (or two) rather than constantly adding to a string.

Most importantly, I don't know that I can properly pull it off without a slightly modified ```lang.getName```.

In summation, I'm learning to swim in the deep end. I think I'm about to make it through, but I can't promise that I will or how long it will take me.  I am having fun learning JS along the way, and I will have no hurt feelings if this doesn't get added to the next version of Quest (if at all).  I am diligently working on it, though.

**Comment by ThePix** [comment]
Okay, no worries, take your time.

**Comment by KVonGit** [comment]
Hello again.

I should have noted that I didn't feel like you were rushing me.  I was doing my best to pleasantly explain that I was sucking at figuring things out.  :p

Now, I <i><b><u>believe</u></b></i> that it might be operational right now, but:

1. I'm sure I have some repetitive code that I need to clean up.
2. I am using a modified ```lang.getName```, but I backed up the original as ```lang.getNameOG```, and it immediately hands everything over to the OG function if ```settings.linksEnabled``` is not set to ```true```.  (I realize this might be an issue, but name modifiers only seem to append the displayed name, whereas I need to replace the displayed name with a link.  (I am probably not grasping another concept here.))
3. I have some ```lang``` things in my actual library file right now.  I plan to move those to another file, but there are three new ones that must exist in order for the item link stuff to work properly:
```js
lang.inside = "inside";
lang.on_top = "on top";
lang.carrying = "carrying";
```

---
I've changed the CSS settings on the dropdown menus.  I wish I could figure out how to make them pop up above the link instead of below it, but that's currently kicking my butt.

Speaking of the CSS, it is now set up via JS at the end of the item_links library.

Test it out:
http://media.textadventures.co.uk/games/R-PODbpzSk_rxIcdAq1RQA/page.html

Check out the code:
https://github.com/KVonGit/Quest6Game1

Or just [the library file](https://github.com/KVonGit/Quest6Game1/blob/main/game/rhfiles/item_links.js).

---
I will clean the code up this evening, making sure I remove any repetitive code I find.

I'll also move the English-specific stuff out of the library file.  I know that definitely doesn't belong in there, but I'd already zipped and uploaded before I remembered I was supposed to fix it.

Also, I think everything with "Redux" at the end has now replaced its original function.  So, I should probably be deleting the originals and losing the "Redux" from the function names.

I'll stop rambling now.  I know that  I'm currently suffering from coding-insanity.  (I'm okay.  Just a little more scatter-brained than usual.)

**Comment by KVonGit** [comment]
Okay . . .  I got rid of the majority of the old code.

These need to exist (I note this in the file, but I feel that is not enough):

```js
lang.inside = "inside";
lang.on_top = "on top";
lang.carrying = "carrying";
```
---
Test it out:
http://media.textadventures.co.uk/games/R-PODbpzSk_rxIcdAq1RQA/page.html

Check out the code:
https://github.com/KVonGit/Quest6Game1

Or just the [library file](https://github.com/KVonGit/Quest6Game1/blob/main/game/rhfiles/item_links.js).

---
I don't exactly know all the ways I should test this.  I realize I haven't tested it with objects with a ```count```.  (Doh!)

I think I made it ```lang``` friendly, but again, I don't exactly know all the ways to test that.

I'm going to take a break for the evening (if I can resist the urge to code).

---
PS

If you click on the version number under the main title when play begins, it will print a link to see the version change file.

![image](https://user-images.githubusercontent.com/30656341/102840307-2d738e00-43c8-11eb-82a2-8d654dbf8980.png)

Of course, you could always simply peruse [versionChanges.js](https://github.com/KVonGit/Quest6Game1/blob/main/game/rhfiles/versionChanges.js).

---
PPS

I'm tempted to close this issue.  If there are issues now, it will be different code than referenced above.

You be the judge, good sir.

Cheers!

**Comment by ThePix** [comment]
> I think I made it lang friendly, but again, I don't exactly know all the ways to test that.

I have the same problem. I do not know what assumptions I am making that do not apply in other languages.

Checked the game and it looks great; I like the double border. Only had a quick look at the code, but for `lang.getNameOG`, you should be able to just do (as long as you do it before you change `lang.getName`):

```
lang.getNameOG = lang.getName
```

**Comment by KVonGit** [comment]
> Only had a quick look at the code, but for lang.getNameOG, you should be able to just do [ . . . ]

Ha ha!  I'll fix that right now.

That is proof of how much I needed to take a break at that moment.  That is normally exactly how I would handle such a thing, were I thinking clearly.  (A good example is the way I defined ```log```.)

Speaking of ```log```, if we set it up the way mrangel showed us on the forum (checking that the ```const``` doesn't exist first), the console shows the wrong line numbers.  Well, technically it shows the line of the code that sets ```log``` using mrangel's method, but that's not the line number I want it to show me.   (You know what I mean (I think).)

I don't know if that will affect anything else negatively, say like if one were to try to use that to set up a change script, but I digress.

---
Today, I am creating something countable to see if I messed up the way countable objects' aliases are printed with the item links.  (Be afraid!  Be very afraid!  Bwahahahaha!)

Seriously, though: I don't trust the way I redefine ```s``` at the end of the modified ```lang.getName```.  I need to test it with pronouns and countables (and probably one or two other things).

**Comment by KVonGit** [comment]
I have added exit links.  It didn't take long. I used the ```cmd``` text processor thingy.

They re-enable when available, too.

Pertex mentioned the item links not re-enabling.  I just figured out how to code it, but do you think it should work that way?  (I currently have it working that way. My only real concern is if it might slow everything down, but I'm thinking it probably won't.  That's just a guess, though.)

**Comment by KVonGit** [comment]
Also, I am 90% sure I need to set it up so whatever getName puts out is used as the displayed text in each item link.  Right now, I don't think I've got every option covered, but I just realized how to easily fix that.

I should have that ready in 2 - 4 (maybe 6) hours, unless I get seriously distracted somehow (which is not very probable).

After I've done that, I really don't know what I might have overlooked.

Will update.

PS

If you test it out on the site, I am currently at version 0.39, which is uploaded, but currently not loading.  (And I know:  Sometimes it takes a short while before it updates the files.)

Cheers!

**Comment by KVonGit** [comment]
Hello!

I *might* be done.  If not, I can't find what I've overlooked.  (I probably missed something, though.)

http://media.textadventures.co.uk/games/R-PODbpzSk_rxIcdAq1RQA/page.html

https://github.com/KVonGit/Quest6Game1/blob/main/game/rhfiles/item_links.js

---
I altered the CSS a little.  I also moved it out of the JS and back into a CSS file.

```css
.exit-link:not(.disabled), .object-link:not(.disabled){
	cursor: pointer;
}
.droplink{
  color:blue;
  font-size: inherit;
}
.droplink:not(.disabled) {
    /*background-color: #3498DB;
    color: blue;
    padding: 16px;
    font-size: 16px;
    border: none;*/
    cursor: pointer;
    text-decoration: underline;
    font-size:inherit;
}

.droplink:hover:not(.disabled), .droplink:focus:not(.disabled), .exit-link:hover:not(.disabled) {
    color: blue;
}

.dropdown {
    position: relative;
    display: inline-block;
}

.dropdown-content {
    display: none;
    position: absolute;
    background-color: #f1f1f1;
    overflow: auto;
	border: 1px solid black;
    box-shadow: 0px 8px 16px 0px rgba(0,0,0,0.2);
    z-index: 1;
}

.dropdown-content span {
    color: black;
    padding: 2px;
    text-decoration: none;
    display: block;
    border: 1px solid black;
}

.dropdown  a:hover {
	background-color: #ddd;
}

.list-link-verb:hover {
	background-color: #ddd;
}

.show {
  display:block;
}
.list-link-verb {
	white-space:nowrap;
}
.exit-link:not(.disabled) {
	text-decoration:underline;
}
```

---
# Things not in the code, but required

These are needed:

```js
lang.inside = "inside";
lang.on_top = "on top";
lang.carrying = "carrying";
```

Changing these two is optional:
```js
lang.contentsForData.surface.prefix = 'on which you see ';
lang.contentsForData.surface.suffix = '';
```

---
This one is not necessary, but I modified this to echo the command clicked:

```js
io.cmdlink = function(command, str) {
  return `<a class="cmd-link" onclick="io.msgInputText('${command}'); parser.parse('${command}');');">${str}</a>`;
};
```

---
I think that's all the info you may need.

As far as I can tell, these links work just like the last release of Quest 5.

---
PS

Are we calling this QuestJS or Quest 6?  Or is the jury still out?

---
PPS

Let me know when you find whatever I've overlooked.

**Comment by ThePix** [comment]
I have optimistically uploaded it and included it in release 0.4., as you have changed the name. It looks good as far as I can see. If we find bugs later, we can correct them in the next update. The rest is like that after all, so do not worry about it too much.

A couple of minor points.

1. This is now in the file:
```
settings.linksEnabled = true
```

If you load the library file, you want them enabled, so it might as well be automatic.

2. The command link I re-wrote with `runCmd` from _util.js (its pretty new, and is there for that exact reason).

**Comment by KVonGit** [comment]
I'm closing this one out (because it's too much work to scroll all the way down here).

Cheers!
